infra/Pos/Slotting/Impl/Util.hs

```diff
@@ -32,21 +37,28 @@ approxSlotUsingOutdated penult t = do
   where
     outdatedEpoch systemStart (Timestamp curTime) epoch EpochSlottingData {..} =
         let duration = convertUnit esdSlotDuration
-            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart) in
+            start = getTimestamp (esdStartDiff `addTimeDiffToTimestamp` systemStart) 
+        in
         unflattenSlotId $
         flattenEpochIndex epoch + fromIntegral ((curTime - start) `div` duration)
 
 -- | Compute current slot from current timestamp based on data
 -- provided by 'MonadSlotsData'.
 slotFromTimestamp
-    :: MonadSlotsData m
-    => Timestamp -> m (Maybe SlotId)
+    :: (MonadSlotsData m, MonadThrow m)
+    => Timestamp 
+    -> m (Maybe SlotId)
 slotFromTimestamp approxCurTime = do
-    SlottingData {..} <- getSlottingData
     systemStart <- getSystemStart
-    let tryEpoch = computeSlotUsingEpoch systemStart approxCurTime
-    let penultRes = tryEpoch sdPenultEpoch sdPenult
-    let lastRes = tryEpoch (succ sdPenultEpoch) sdLast
+    sdLast      <- getLastEpochSDEmpatically
+    sdPenult    <- getPenultEpochSDEmpatically
+
+    lastEpochI  <- getEpochLastIndex 
+    let sdPenultEpoch = lastEpochI - 1
+
+    let penultRes = computeSlotUsingEpoch systemStart approxCurTime sdPenultEpoch sdPenult
+    let lastRes   = computeSlotUsingEpoch systemStart approxCurTime (succ lastEpochI) sdLast
+
     return $ penultRes <|> lastRes
```

Comment:
Hm, this code should be reworked. Currently it checks only the last two epochs. So it will return `Nothing` for an old epoch. But the task is to improve it! It should find an appropriate epoch and use its slotting data.
If I recall correctly, there was right logic in the original PR.
It should return `Nothing` only for timestamps which are after the last epoch from `SlottingData`.